compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2DToImg2Col.cpp

```diff
@@ -375,6 +375,155 @@ class ConvertDepthwiseConv2DNhwcHwc final
+
+    auto nloops = colTensorShape.size();
+
+    auto parallel = utils::IteratorType::parallel;
+    auto reduction = utils::IteratorType::reduction;
+    SmallVector<utils::IteratorType, 3> img2colIterators(nloops, parallel);
+
+    SmallVector<AffineMap, 4> img2colIndexingMaps = {
+        AffineMap::get(nloops, 0, inputExprs, rewriter.getContext()),
+        AffineMap::getMultiDimIdentityMap(nloops, rewriter.getContext())};
+
+    auto img2ColTensor = rewriter.create<linalg::GenericOp>(
+        loc, colTensor.getType(),
+        /*inputs=*/input, /*outputs=*/colTensor, img2colIndexingMaps,
+        img2colIterators,
+        [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+          nestedBuilder.create<linalg::YieldOp>(nestedLoc, args[0]);
+        });
+
+    SmallVector<ReassociationIndices> filterReassocIndices = {{0}, {1, 2, 3}};
+    auto reshapedFilterType =
+        RankedTensorType::get({oc, fh * fw * ic}, inputType.getElementType());
+    Value reshapedFilter = rewriter.create<tensor::CollapseShapeOp>(
+        loc, reshapedFilterType, filter, filterReassocIndices);
+
+    SmallVector<ReassociationIndices> img2ColTensorReassocIndices;
+    SmallVector<ReassociationIndices> outputReassocIndices;
+    RankedTensorType reshapedImg2ColTensorType, reshapedOutputType;
+    if (n == 1) {
+      img2ColTensorReassocIndices = {{0, 1, 2, 3}, {4, 5}};
+      outputReassocIndices = {{0, 1}, {2, 3}};
+
+      reshapedImg2ColTensorType = RankedTensorType::get(
+          {fh * fw * ic, oh * ow}, inputType.getElementType());
+      reshapedOutputType =
+          RankedTensorType::get({oc, oh * ow}, outputType.getElementType());
+    } else {
+      img2ColTensorReassocIndices = {{0}, {1, 2, 3}, {4, 5}};
+      outputReassocIndices = {{0}, {1}, {2, 3}};
+
+      reshapedImg2ColTensorType = RankedTensorType::get(
+          {n, fh * fw * ic, oh * ow}, inputType.getElementType());
+      reshapedOutputType =
+          RankedTensorType::get({n, oc, oh * ow}, outputType.getElementType());
+    }
+
+    Value reshapedImg2ColTensor = rewriter.create<tensor::CollapseShapeOp>(
+        loc, reshapedImg2ColTensorType, img2ColTensor.getResult(0),
+        img2ColTensorReassocIndices);
+
+    Value reshapedOutput = rewriter.create<tensor::CollapseShapeOp>(
+        loc, reshapedOutputType, output, outputReassocIndices);
+
+    Value result;
+    if (n == 1) {
+      auto matmulOp = rewriter.create<linalg::MatmulOp>(
+          loc, reshapedOutputType,
+          ArrayRef<Value>{reshapedFilter, reshapedImg2ColTensor},
+          ArrayRef<Value>{reshapedOutput});
+      result = matmulOp.getResults().front();
+    } else {
+      // For cases where batch is not 1, we need to keep the batch dimension
+      // separate. However the batch dimension is only used in indexing the
+      // input and output. So we cannot use existing linalg named ops like
+      // linalg.batch_matmul; doing it with a linalg.generic instead.
```

Comment:
Can you explain this a bit more, its not immediately clear to me.

Reply:
This is the same as for nhwc, but in short the column feature matrix will have a batch dimension here while the filter won't so none of the named ops will work (e.g. MxK * (Bx)KxN).

I can update the comment here.